src/streaming/zarr.stream.cpp

```diff
@@ -0,0 +1,494 @@
+#include "macros.hh"
+#include "zarr.stream.hh"
+#include "acquire.zarr.h"
+
+#include <blosc.h>
+
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace {
+bool
+is_s3_acquisition(const struct ZarrStreamSettings_s* settings)
+{
+    return nullptr != settings->s3_settings;
+}
+
+bool
+is_compressed_acquisition(const struct ZarrStreamSettings_s* settings)
+{
+    return nullptr != settings->compression_settings;
+}
+
+[[nodiscard]]
+std::string
+trim(const char* s)
+{
+    if (!s || !*s) {
+        return {};
+    }
+
+    const size_t length = strlen(s);
+
+    // trim left
+    std::string trimmed(s, length);
+    trimmed.erase(trimmed.begin(),
+                  std::find_if(trimmed.begin(), trimmed.end(), [](char c) {
+                      return !std::isspace(c);
+                  }));
+
+    // trim right
+    trimmed.erase(std::find_if(trimmed.rbegin(),
+                               trimmed.rend(),
+                               [](char c) { return !std::isspace(c); })
+                    .base(),
+                  trimmed.end());
+
+    return trimmed;
+}
+
+[[nodiscard]]
+bool
+validate_s3_settings(const ZarrS3Settings* settings)
+{
+    std::string trimmed = trim(settings->endpoint);
+    if (trimmed.empty()) {
+        LOG_ERROR("S3 endpoint is empty");
+        return false;
+    }
```

Comment:
The emptiness check is repeated multiple times. We should extract it into a separate function that takes an untrimmed string and an error message as parameters.
```cpp
if (is_empty_string(settings->endpoint, "S3 endpoint is empty")) {
    return false;
}
```
This is also a good candidate for a utility function.